### ISSUE TITLE
Update databricks provider to use TriggerOperator

### ISSUE THREAD

**Issue by chinwobble** [apache/airflow#18999]
### Description

The databricks provider can be updated to a deferrable task introduced in airflow 2.2.0. This will significnatly reduce cpu usage, memory usage on LocalExecutor and improve reliability since state is stored in the airflow metastore.

### Use case/motivation

Currently the databricks operators works by calling the databricks REST API to submit a spark job and polling it every 20 secs to check when it is done. This creates the following inefficiencies:
- If the airflow executor process crashes, a duplicate job can be created in databricks since airflow doesn't save the databricks job run id.
- Each task instance of the databricks operators run its own process, however 95% of the time the process is idle and waiting to repoll the databricks API. If you want to 50 jobs in parallel, you will use a non trivial amount of memory / CPU just to poll a rest API.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by eskarimov** [comment]
@chinwobble thank you so much, it's a good base to start working on! 👍

Regarding the task duration - I haven't checked yet, but I thought it's implemented so that execution time counted over total time, so an operator might fail even while it's deferred.

**Comment by chinwobble** [comment]
> 
> 
> @chinwobble thank you so much, it's a good base to start working on! 👍
> 
> Regarding the task duration - I haven't checked yet, but I thought it's implemented so that execution time counted over total time, so an operator might fail even while it's deferred.

I've checked the task duration.
If your task defers, that deferral time is not considered as part of the task duration.

**Comment by eskarimov** [comment]
@chinwobble you're right, I see the same behaviour... I've spotted though when I don't pass `timeout` argument to `self.defer()`, a deferred task will fail if its runtime more than `execution_timeout` set for the Operator calling it. 
I.e. timeout for a deferred task is equal to execution timeout for an Operator in this case. [Here is the code behind](https://github.com/apache/airflow/blob/94a0a0e8ce4d2b54cd6a08301684e299ca3c36cb/airflow/models/taskinstance.py#L1547-L1555)

Also we might found a a bug, because I see [a code part in `TaskInstance`](https://github.com/apache/airflow/blob/94a0a0e8ce4d2b54cd6a08301684e299ca3c36cb/airflow/models/taskinstance.py#L1495-L1500) trying to check for timeout when task executed after deferred state. The issue is that `self.start_date` is changing every time when a task is continued after the deferred state.

**Comment by chinwobble** [comment]
> runtime

![image](https://user-images.githubusercontent.com/18226001/139786606-5a718625-1a8b-4db6-82d3-4f8cd416a3e6.png)

**Comment by chinwobble** [comment]
We are using localExecutor and it seems like the deferral consumes an unnecessary amount of tasks.
When implementing deferral, it looks like this is what is happening.
The way deferral work is like this:

 1. schedulerJob decides a task is ready to be queued
 2. Queued jobs run on a local executor and start a new job consuming 1 worker slot.
 3. The execute method is started, it takes 1 sec to submit the job to databricks
 4. self.defer() is run and queues a trigger.
 5. This throws an exception and causes the task job to terminate and the worker slot is returned to the pool
 6. The triggerer is a separate process that can asynchronously process triggers and determine when to a resume a task
 7. The trigger yields a result causing the scheduler to requeue the original databricks task possibly on a different method
 8. A new executor worker process (second job start) is started (consuming 1 worker slot), when complete it marks the task as succeeded or failed
 9. The task is complete and all workers are returned to the pool.

**Comment by eskarimov** [comment]
FYI just wanted to share that I'm still willing to introduce deferrable operators for Databricks, and working on the code.
Hope to create a PR later this week.
There're a couple of moments I'd like to discuss with someone more experienced than me, @chinwobble I'd be very happy if you review the code, will add you to the reviewers list if you don't mind.

Regarding the timeout - I've created a separate issue #19382 to investigate and fix it, shared the findings there.
@chinwobble could I add your comment above there into the issue description? It's a nice summary on how it works internally

**Comment by chinwobble** [comment]
@eskarimov Happy to review your code. I'm not the best python dev, so I think I could learn a thing or two from you.

https://github.com/apache/airflow/issues/18999#issuecomment-958449899

These deferrable operators are a good first step in making airflow more scalable. 
However, there are a few issues that make me feel its not ready yet:
- task duration during deferral is not counted
- for the databricks use case, there is an unnecessary number of processes being started / killed 
- More polling is introduced, for our pipeline that takes 2.5 hours / day, an extra 20 mins was added after implementing deferrable operators.

**Comment by chinwobble** [comment]
@eskarimov  nice work getting this one merged!
